contrib/storage-splunk/src/main/java/org/apache/drill/exec/store/splunk/SplunkBatchWriter.java

```diff
@@ -0,0 +1,308 @@
+import org.apache.drill.exec.record.VectorAccessible;
+import org.apache.drill.exec.store.AbstractRecordWriter;
+import org.apache.drill.exec.store.EventBasedRecordWriter.FieldConverter;
+import org.apache.drill.exec.vector.complex.reader.FieldReader;
+import org.json.simple.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class SplunkBatchWriter extends AbstractRecordWriter {
+
+  private static final Logger logger = LoggerFactory.getLogger(SplunkBatchWriter.class);
+  private static final String DEFAULT_SOURCETYPE = "drill";
+  private final UserCredentials userCredentials;
+  private final List<String> tableIdentifier;
+  private final SplunkWriter config;
+  private final Args eventArgs;
+  protected final Service splunkService;
+  private JSONObject splunkEvent;
+  protected Index destinationIndex;
+
+
+  public SplunkBatchWriter(UserCredentials userCredentials, List<String> tableIdentifier, SplunkWriter config) {
+    this.config = config;
+    this.tableIdentifier = tableIdentifier;
+    this.userCredentials = userCredentials;
+
+    SplunkConnection connection = new SplunkConnection(config.getPluginConfig(), userCredentials.getUserName());
+    this.splunkService = connection.connect();
+
+    // Populate event arguments
+    this.eventArgs = new Args();
+    eventArgs.put("sourcetype", DEFAULT_SOURCETYPE);
+  }
+
+  @Override
+  public void init(Map<String, String> writerOptions) throws IOException {
+    // No op
+  }
+
+  /**
+   * Update the schema in RecordWriter. Called at least once before starting writing the records. In this case,
+   * we add the index to Splunk here. Splunk's API is a little sparse and doesn't really do much in the way
+   * of error checking or providing feedback if the operation fails.
+   *
+   * @param batch {@link VectorAccessible} The incoming batch
+   */
+  @Override
+  public void updateSchema(VectorAccessible batch) {
+    logger.debug("Updating schema for Splunk");
+
+    //Get the collection of indexes
+    IndexCollection indexes = splunkService.getIndexes();
+    try {
+      String indexName = tableIdentifier.get(0);
+      indexes.create(indexName);
+      destinationIndex = splunkService.getIndexes().get(indexName);
+    } catch (Exception e) {
+      // We have to catch a generic exception here, as Splunk's SDK does not really provide any kind of
+      // failure messaging.
+      throw UserException.systemError(e)
+        .message("Error creating new index in Splunk plugin: " + e.getMessage())
+        .build(logger);
+    }
+  }
+
+
+  @Override
+  public void startRecord() {
+    logger.debug("Starting record");
+    splunkEvent = new JSONObject();
+  }
+
+  @Override
+  public void endRecord() throws IOException {
+    logger.debug("Ending record");
+    // Write the event to the Splunk index
+    destinationIndex.submit(eventArgs, splunkEvent.toJSONString());
```

Comment:
I found the following on [this Splunk doc page](https://dev.splunk.com/enterprise/docs/devtools/java/sdk-java/howtousesdkjava/howtogetdatasdkjava), can we take advantage of the 'attach' method that they suggest?

> The submit method opens a new connection for each event, so if you have many events to send, this method may not perform well. In this case, use the attach method get an open socket to Splunk Enterprise that you can write your events to. Here is an example of sending an event (with a timestamp) to a socket opened by the attach method

Reply:
@jnturton 
I actually tried this first and I couldn't get Splunk to actually write any data.  I literally cut/pasted their code into Drill to no avail.

Reply:
I think there may be some bug in the Splunk SDK.

Reply:
@cgivre can we leave a comment explaining this to readers then?

Reply:
@jnturton I figured this out.   Using Splunk's sample code from their SDK documentation resulted in Splunk not parsing the fields correctly which broke all the unit tests, and didn't work.  I did some experiments and found that removing the date actually solved the issue.

Splunk's SDK provides a method for writing to a socket which does all the error handling.  I used that because that was what the docs recommended, however that method does not allow you to set some of the properties that the other insert methods do.  But I'm not debugging Splunk's SDK for free.